libs/ml/src/utilities/graph_builder.cpp

```diff
@@ -0,0 +1,563 @@
+#include "ml/ops/activations/sigmoid.hpp"
+#include "ml/ops/activations/softmax.hpp"
+#include "ml/ops/add.hpp"
+#include "ml/ops/avg_pool_1d.hpp"
+#include "ml/ops/avg_pool_2d.hpp"
+#include "ml/ops/concatenate.hpp"
+#include "ml/ops/convolution_1d.hpp"
+#include "ml/ops/convolution_2d.hpp"
+#include "ml/ops/divide.hpp"
+#include "ml/ops/embeddings.hpp"
+#include "ml/ops/exp.hpp"
+#include "ml/ops/flatten.hpp"
+#include "ml/ops/log.hpp"
+#include "ml/ops/loss_functions/cross_entropy_loss.hpp"
+#include "ml/ops/loss_functions/mean_square_error_loss.hpp"
+#include "ml/ops/loss_functions/softmax_cross_entropy_loss.hpp"
+#include "ml/ops/mask_fill.hpp"
+#include "ml/ops/matrix_multiply.hpp"
+#include "ml/ops/max_pool.hpp"
+#include "ml/ops/max_pool_1d.hpp"
+#include "ml/ops/max_pool_2d.hpp"
+#include "ml/ops/maximum.hpp"
+#include "ml/ops/multiply.hpp"
+#include "ml/ops/one_hot.hpp"
+#include "ml/ops/placeholder.hpp"
+#include "ml/ops/reduce_mean.hpp"
+#include "ml/ops/reshape.hpp"
+#include "ml/ops/slice.hpp"
+#include "ml/ops/sqrt.hpp"
+#include "ml/ops/squeeze.hpp"
+#include "ml/ops/strided_slice.hpp"
+#include "ml/ops/subtract.hpp"
+#include "ml/ops/switch.hpp"
+#include "ml/ops/tanh.hpp"
+#include "ml/ops/top_k.hpp"
+#include "ml/ops/transpose.hpp"
+#include "ml/ops/weights.hpp"
+#include "ml/saveparams/saveable_params.hpp"
+
+namespace fetch {
+namespace ml {
+namespace utilities {
+
+template <typename T>
+void BuildNodeAndInsertTrainables(NodeSaveableParams<T> const &nsp, std::string const &name,
+                                  std::shared_ptr<Graph<T>> g)
+{
+  auto node = std::make_shared<Node<T>>();
+  // construct the concrete op
+  // if the op is a layer, invoke the layer Build method
+
+  std::shared_ptr<ops::Ops<T>> op_ptr;
+  switch (nsp.operation_type)
+  {
+  case ops::Abs<T>::OpCode():
+  {
+    op_ptr = GetOp<ops::Abs<T>>(nsp.op_save_params);
+    node->SetNodeSaveableParams(nsp, op_ptr);
+    g->AddTrainable(node, name);
+    break;
+  }
+  case ops::Add<T>::OpCode():
+  {
+    op_ptr = GetOp<ops::Add<T>>(nsp.op_save_params);
+    node->SetNodeSaveableParams(nsp, op_ptr);
+    g->AddTrainable(node, name);
+    break;
+  }
+  case ops::Concatenate<T>::OpCode():
+  {
+    op_ptr = GetOp<ops::Concatenate<T>>(nsp.op_save_params);
+    node->SetNodeSaveableParams(nsp, op_ptr);
+    g->AddTrainable(node, name);
+    break;
+  }
+  case ops::Constant<T>::OpCode():
+  {
+    op_ptr = GetOp<ops::Constant<T>>(nsp.op_save_params);
+    node->SetNodeSaveableParams(nsp, op_ptr);
+    g->AddTrainable(node, name);
+    break;
+  }
+  case ops::Convolution1D<T>::OpCode():
+  {
+    op_ptr = GetOp<ops::Convolution1D<T>>(nsp.op_save_params);
+    node->SetNodeSaveableParams(nsp, op_ptr);
+    g->AddTrainable(node, name);
+    break;
+  }
+  case ops::Convolution2D<T>::OpCode():
+  {
+    op_ptr = GetOp<ops::Convolution2D<T>>(nsp.op_save_params);
+    node->SetNodeSaveableParams(nsp, op_ptr);
+    g->AddTrainable(node, name);
+    break;
```

Comment:
pls pay attention, these

    node->SetNodeSaveableParams(nsp, op_ptr);
    g->AddTrainable(node, name);

are just copied to each case, so I guess we can move them outside of the switch at all. The only cases that do not contatin these two lines are cases that throw exceptions.

Reply:
done